File: gateway/webui/src/components/settings/panes/tool-permission-patch.ts
// gateway/webui/src/components/settings/panes/tool-permission-patch.ts
//
// Pure helpers shared by every editable row in the Tools pane. Extracted so
// the invariants this feature depends on have their own tests, independent
// of rendering: see `tool-permission-patch.test.ts`.
//
// WHY THE MASTER CONTROL CANNOT JUST WRITE THE WILDCARD. Every account is
// seeded with a NAMED permission entry for every catalog tool its role can
// execute (`gateway/src/profile-store/profile-defaults.ts#applyProfileDefaults`,
// called at account creation). `resolve-tool-permission.ts`'s
// `storedPermissionFor` reads a tool's own named key BEFORE the server's
// wildcard (`perServer[toolName] ?? perServer[ALL_TOOLS_PERMISSION_KEY]`) —
// so for a normal, already-seeded account, EVERY tool already has an answer
// that outranks the wildcard, and writing only `permissions[server]["*"]`
// changes nothing for any of them. The wildcard only ever governs a tool
// with NO named entry: one the operator adds to the catalog after this
// account was seeded, or one a role change newly grants (a promoted/demoted
// account's table is not re-seeded). `withServerMasterPermission` below is
// what makes the control actually hide/restore an already-named tool too.
import type { ToolPermission, ToolPermissionOrClear, ToolPermissionPatchMap } from "@sentient/config";
import type { McpToolView } from "../../../services/profile-api.js";

/**
 * Returns a NEW permissions map with exactly (serverId, toolName) set to
 * `permission`. Every other server, and every other tool-name key already
 * under `serverId` (including a previously-set wildcard entry), is carried
 * forward untouched.
 *
 * `permission` MAY be `null` — a CLEAR, not a delete. Passed straight
 * through to the PUT body: the gateway's merge (`profile-update.ts`) is what
 * turns a `null` leaf into an absent key server-side, so a person can return
 * a tool (or a server's wildcard) to "no stored opinion, let the role
 * template answer" without this client ever deleting a JS object key itself.
 *
 * NEVER DELETES A KEY ON THIS SIDE EITHER — but not for the reason this
 * comment used to give. Dropping a SERVER key from the body is simply a NO-OP
 * for that server, NOT the "off forever" the absent-server rule describes.
 * They are different maps: `profile-update.ts` does a per-server, per-tool
 * DELTA merge, so a key the body omits is left exactly as stored, while the
 * absent-server rule (`resolve-tool-permission.ts`'s `storedPermissionFor`)
 * is a property of the STORED table — a server with no key there at all — and
 * no PUT can reach that state by omission.
 *
 * Keys are carried forward to match the full-resend convention every other
 * PUT field follows. Building the next map by spreading the PREVIOUS one
 * (rather than reconstructing it from the catalog view) is what keeps a
 * server the catalog excluded for an unrelated reason — stdio transport, zero
 * role-governable tools — represented in the body at all, so this person's
 * saved table stays a complete statement of what they chose.
 */
export function withToolPermission(
  permissions: ToolPermissionPatchMap | undefined,
  serverId: string,
  toolName: string,
  permission: ToolPermissionOrClear,
): ToolPermissionPatchMap {
  return {
    ...permissions,
    [serverId]: { ...permissions?.[serverId], [toolName]: permission },
  };
}

/**
 * The server master control's actual write: a NAMED value for every catalog
 * tool this role can govern (`toolNames`, the role-narrowed list the pane
 * already renders — `entry.tools.map(t => t.name)`), PLUS the wildcard for
 * whatever that list cannot enumerate (see the module comment above for why
 * the wildcard alone is a no-op on a normal, already-seeded account).
 *
 * "Off" writes an explicit `"off"` everywhere — a real, stored opinion that
 * hides every one of this server's tools from the model.
 *
 * "On" writes an explicit `null` everywhere instead of a concrete value:
 * each named tool falls back to ITS OWN role-template answer rather than
 * one blanket value, which is what keeps a `confirm`-tier tool's `ask` from
 * becoming an auto-approved `allow` (the bug this whole clearing mechanism
 * exists to close — see `withToolPermission`'s doc comment).
 *
 * "ON" ALSO ERASES ANY PER-TOOL OVERRIDE THE PERSON SET DELIBERATELY ON THIS
 * SERVER. The stored table records no provenance — a value seeded at
 * account creation and a value a person typed into this exact tool's own
 * Select five minutes ago are the same kind of key, indistinguishable once
 * written. There is no way to clear "everything the template already
 * agreed with" while preserving "everything the person chose on purpose"
 * without inventing data the table does not carry, which is why this
 * control is framed as "Hide all / Reset to role defaults" in the pane's
 * copy, not "Hide all / Undo my last change".
 */
export function withServerMasterPermission(
  permissions: ToolPermissionPatchMap | undefined,
  serverId: string,
  toolNames: readonly string[],
  wildcardKey: string,
  turnOn: boolean,
): ToolPermissionPatchMap {
  const value: ToolPermissionOrClear = turnOn ? null : "off";
  const namedWrites = Object.fromEntries(toolNames.map((name) => [name, value]));
  return {
    ...permissions,
    [serverId]: { ...permissions?.[serverId], ...namedWrites, [wildcardKey]: value },
  };
}

/**
 * What a tool's Select should show: this person's own pending or
 * previously-saved edit for that EXACT tool name if there is one, else the
 * catalog's already-resolved value. The catalog's `tool.permission` was
 * computed server-side by `resolve-tool-permission.ts` from the saved
 * profile plus the role template — this deliberately does not re-simulate
 * that cascade (wildcard, role template, fail-closed backstop) client-side,
 * so the settings screen can never drift from what the ToolBroker actually
 * enforces. One consequence: a PENDING (unsaved) master-control write is not
 * reflected in an individual tool's row until the profile is saved and the
 * catalog is refetched — consistent with this pane's own "changes apply
 * after you save settings" copy.
 *
 * A per-tool Select never WRITES `null` (it only ever offers the four real
 * permissions), but `??` treats a stray `null` the same as "no opinion" here
 * too — the honest fallback if one were ever present — which is also why
 * this keeps returning a concrete `ToolPermission`, never `null`.
 */
export function effectiveToolPermission(
  permissions: ToolPermissionPatchMap | undefined,
  serverId: string,
  tool: McpToolView,
): ToolPermission {
  return permissions?.[serverId]?.[tool.name] ?? tool.permission;
}

/**
 * What a server's master control should show: this person's own pending or
 * previously-saved wildcard write if there is one, else the catalog's
 * snapshot of it at load time (`McpCatalogEntryView.wildcardPermission`).
 *
 * DELIBERATELY NOT `??`. A pending CLEAR is stored as a real, own `null`
 * entry — `permissions?.[serverId]?.[wildcardKey] ?? catalogWildcard` would
 * treat that `null` as "nothing here" and silently fall back to the catalog's
 * (possibly stale, possibly `"off"`) snapshot, which is exactly backwards:
 * the person just asked to clear it. `Object.hasOwn` is what tells "an
 * explicit pending edit exists, and it happens to be null" apart from "no
 * edit was made here at all".
 */
export function effectiveWildcardPermission(
  permissions: ToolPermissionPatchMap | undefined,
  serverId: string,
  wildcardKey: string,
  catalogWildcard: ToolPermission | null,
): ToolPermission | null {
  const serverMap = permissions?.[serverId];
  if (serverMap && Object.hasOwn(serverMap, wildcardKey)) {
    // `noUncheckedIndexedAccess` types this read as possibly `undefined`
    // even though `Object.hasOwn` just proved the key exists. An explicit
    // `=== undefined` check (never `??`) is what lets a REAL `null` pass
    // through untouched — `??` cannot tell "clear" apart from "absent".
    const value = serverMap[wildcardKey];
    return value === undefined ? catalogWildcard : value;
  }
  return catalogWildcard;
}

File: shared/mobile-sdk/src/commonMain/kotlin/io/sentient/mobilesdk/settings/ToolPermissionPatch.kt
// ---------------------------------------------------------------------------
// ToolPermissionPatch — pure helpers shared by every tool-permission control on
// every platform (Android + iOS Tools screens). Ported 1:1 from gateway/webui's
// src/components/settings/panes/tool-permission-patch.ts — mirror THAT file's
// semantics and its own doc comments carry the fuller reasoning; this file only
// restates what a Kotlin caller needs to know to use it safely.
//
// WHY A SERVER MASTER CONTROL CANNOT JUST WRITE THE WILDCARD. Every account is
// seeded with a NAMED permission entry for every catalog tool its role can
// execute (gateway profile-defaults.ts#applyProfileDefaults). The resolver
// reads a tool's own named key BEFORE the server's wildcard — so on an
// already-seeded account, every tool already has an answer that outranks the
// wildcard, and writing only the wildcard changes nothing for any of them.
// [withServerMasterPermission] is what makes a bulk write actually reach an
// already-named tool too.
// ---------------------------------------------------------------------------
package io.sentient.mobilesdk.settings

/**
 * Returns a NEW permissions map with exactly ([serverId], [toolName]) set to
 * [permission]. Every other server, and every other tool-name key already
 * under [serverId] (including a previously-set wildcard entry), is carried
 * forward untouched.
 *
 * [permission] MAY be `null` — a CLEAR, not a delete. The gateway's merge is
 * what turns a `null` leaf into an absent stored key; this function (and every
 * caller of it) never deletes a map key itself.
 *
 * DROPPING A SERVER KEY FROM THE PATCH IS SIMPLY A NO-OP for that server, and
 * NOT the "off forever" the absent-server rule describes. The two are different
 * maps. `profile-update.ts` does a per-server, per-tool DELTA merge: a key this
 * patch omits is left exactly as stored. The absent-server rule
 * (`resolve-tool-permission.ts#storedPermissionFor`) is a property of the
 * STORED table — a server that has no key there at all — and no PUT can put the
 * table into that state by omission. Keys are carried forward here to match the
 * full-resend convention every other PUT field follows (see
 * [mergeToolPermissionPatch], which states the same thing), not to avert a
 * wipe.
 */
fun withToolPermission(
    permissions: ToolPermissionPatchMap?,
    serverId: String,
    toolName: String,
    permission: ToolPermission?,
): ToolPermissionPatchMap {
    val serverMap = permissions?.get(serverId).orEmpty()
    return permissions.orEmpty() + (serverId to (serverMap + (toolName to permission)))
}

/**
 * The server master control's actual write: a NAMED value for every
 * role-governable catalog tool on this server ([toolNames]), PLUS the
 * wildcard for whatever that list cannot enumerate (see this file's header
 * for why the wildcard alone is a no-op on a normal, already-seeded account).
 *
 * [turnOn] = `false` ("off"): writes an explicit [ToolPermission.OFF]
 * everywhere — a real, stored opinion that hides every one of this server's
 * tools from the model.
 *
 * [turnOn] = `true` ("on"): writes an explicit `null` everywhere instead of a
 * concrete value — each named tool falls back to ITS OWN role-template answer
 * rather than one blanket value, which is what keeps a `confirm`-tier tool's
 * ASK from becoming an auto-approved ALLOW (a shipped-and-caught-in-review
 * privilege-escalation bug on web; do not re-introduce it here). "On" ALSO
 * erases any per-tool override the person set deliberately on this server —
 * the stored table records no provenance, so there is no way to clear
 * "everything the template already agreed with" while preserving "everything
 * the person chose on purpose" without inventing data the table does not
 * carry.
 */
fun withServerMasterPermission(
    permissions: ToolPermissionPatchMap?,
    serverId: String,
    toolNames: List<String>,
    wildcardKey: String,
    turnOn: Boolean,
): ToolPermissionPatchMap {
    val value: ToolPermission? = if (turnOn) null else ToolPermission.OFF
    val namedWrites = toolNames.associateWith { value }
    val serverMap = permissions?.get(serverId).orEmpty()
    return permissions.orEmpty() + (serverId to (serverMap + namedWrites + (wildcardKey to value)))
}

/**
 * What a tool's permission control should show: this person's own pending or
 * previously-saved edit for that EXACT tool name if there is one, else the
 * catalog's already-resolved [McpToolView.permission]. Deliberately does NOT
 * re-simulate the resolver's cascade (wildcard, role template, fail-closed
 * backstop) client-side, so the settings screen can never drift from what the
 * ToolBroker actually enforces.
 *
 * A per-tool control never WRITES `null` (it only ever offers the four real
 * permissions), but a stray pending `null` reads as "no opinion" here too —
 * the honest fallback if one were ever present — which is also why this
 * always returns a concrete [ToolPermission], never `null`.
 */
fun effectiveToolPermission(
    permissions: ToolPermissionPatchMap?,
    serverId: String,
    tool: McpToolView,
): ToolPermission = permissions?.get(serverId)?.get(tool.name) ?: tool.permission

/**
 * What a server's master control should show: this person's own pending or
 * previously-saved wildcard write if there is one, else the catalog's
 * snapshot of it at load time ([McpCatalogEntry.wildcardPermission]).
 *
 * DELIBERATELY NOT a plain `?:` fallback. A pending CLEAR is stored as a real,
 * own `null` entry — falling back to the catalog snapshot whenever the read
 * value is `null` would treat that clear as "nothing here" and silently
 * resurrect a stale (possibly OFF) snapshot, exactly backwards from what the
 * person just asked for. [Map.containsKey] is what tells "an explicit pending
 * edit exists, and it happens to be null" apart from "no edit was made here".
 */
fun effectiveWildcardPermission(
    permissions: ToolPermissionPatchMap?,
    serverId: String,
    wildcardKey: String,
    catalogWildcard: ToolPermission?,
): ToolPermission? {
    val serverMap = permissions?.get(serverId) ?: return catalogWildcard
    if (!serverMap.containsKey(wildcardKey)) return catalogWildcard
    return serverMap[wildcardKey]
}

/** Product-group spellings for new mobile settings callers. */
fun withProductGroupToolPermission(
    permissions: ToolPermissionPatchMap?,
    groupId: String,
    toolName: String,
    permission: ToolPermission?,
): ToolPermissionPatchMap = withToolPermission(permissions, groupId, toolName, permission)

fun withProductGroupMasterPermission(
    permissions: ToolPermissionPatchMap?,
    groupId: String,
    toolNames: List<String>,
    wildcardKey: String,
    turnOn: Boolean,
): ToolPermissionPatchMap = withServerMasterPermission(permissions, groupId, toolNames, wildcardKey, turnOn)

/**
 * Merges a session's accumulated pending edits ([overlay] — built incrementally via
 * [withToolPermission] / [withServerMasterPermission], seeded from an empty map) on top of
 * the ORIGINALLY-LOADED stored permissions ([base]) at SAVE time, producing the patch to
 * actually PUT.
 *
 * NOT NEEDED TO KEEP AN UNTOUCHED SERVER FROM TURNING OFF — the gateway's PUT handler
 * (`gateway/src/profile-store/profile-update.ts`) does a genuine per-server, per-tool DELTA
 * merge: a server or tool key the incoming patch omits is left UNTOUCHED in storage, never
 * reinterpreted as "off" because this one request didn't mention it. Sending [overlay] alone
 * (only the keys this session actually touched) would be safe server-side. This function
 * exists instead to match the FULL-RESEND convention every other [ProfileV1PutBody] field
 * already follows (model/voice/audio/persona/compression/advanced are always sent as the
 * complete current value, never a partial diff) — without it, `tools.permissions` would be
 * the one field in the body that behaves like a true partial patch, inconsistent with the
 * rest; the webui's own tools-pane follows the identical full-resend convention by keeping
 * the FULL loaded permissions as its live draft state rather than a separate overlay.
 *
 * This merges at the TOOL-NAME level within each server too (not a per-server whole-map
 * replace): [base]'s other tools under a server the person partially edited are carried
 * forward exactly as [withToolPermission] already promises for a single edit — this is just
 * that same carry-forward guarantee applied once, across every server, at the point a whole
 * session's edits are flattened into one write.
 *
 * A concrete [base] leaf is always a valid patch leaf (see [ProfileV1.toPutBody]), so only
 * [overlay]'s own keys can carry a `null` clear — this function never invents one.
 */
fun mergeToolPermissionPatch(base: ToolPermissionMap?, overlay: ToolPermissionPatchMap): ToolPermissionPatchMap {
    val servers = base.orEmpty().keys + overlay.keys
    return servers.associateWith { server ->
        val baseServerMap: Map<String, ToolPermission?> = base?.get(server).orEmpty()
        baseServerMap + overlay[server].orEmpty()
    }
}
